### ISSUE TITLE
Can energy-preserving callbacks be propagated into variational_integrator ?

### ISSUE THREAD

**Issue by Datseris** [JuliaDynamics/DynamicalSystemsBase.jl#8]
This would make `lyapunovs` and `gali` also conserve energy.

**Comment by Datseris** [comment]
to make the discussion easier, here is the source code of `variational_integrator` for version 0.3.1 : 
```julia
function variational_integrator(ds::ContinuousDS, k::Int, T,
    S::AbstractMatrix; diff_eq_kwargs = Dict())

    f! = ds.prob.f
    jac! = ds.jacob!
    J = ds.J
    # the equations of motion `veom!` evolve the system and
    # k deviation vectors. Notice that the k deviation vectors
    # can also be considered a D×k matrix (which is the case
    # at `lyapunovs` function).
    # The e.o.m. for the system is f!(t, u , du).
    # The e.o.m. for the deviation vectors (tangent dynamics) are simply:
    # dY/dt = J(u) ⋅ Y
    # with J the Jacobian of the vector field at the current state
    # and Y being each of the k deviation vectors
    veom! = (t, u, du) -> begin
        us = view(u, :, 1)
        f!(t, us, view(du, :, 1))
        jac!(t, us, J)
        A_mul_B!(view(du, :, 2:k+1), J, view(u, :, 2:k+1))
    end

    if typeof(T) <: Real
        varprob = ODEProblem{true}(veom!, S, (zero(T), T))
    else
        varprob = ODEProblem{true}(veom!, S, T)
    end

    solver, newkw = get_solver(diff_eq_kwargs)
    vintegrator = init(varprob, solver; newkw..., save_everystep=false)
    return vintegrator
end
```

**Comment by Datseris** [comment]
Closed in favor of #4
